idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,968 @@
+        day = 1
+
+        if j > len(raw_data):
+            raise ValueError(
+                f"tried to generate daily data, but horizon length of {self._horizon_length} exceeds raw_data length of {len(raw_data)}"
+            )
+
+        while j <= len(raw_data):
+            daily_data[day] = raw_data[i:j].reset_index(drop=True)
+            i = j
+            j = j + self._horizon_length
+            day = day + 1
+
+        return daily_data
+
+    def get_optimal_n_clusters(
+        self,
+        daily_data,
+        kmin=None,
+        kmax=None,
+        plot=False,
+    ):
+        """
+        Determines the appropriate number of clusters needed for a
+        given price signal.
+
+        Args:
+            daily_data: LMP signal grouped by days (output of generate_daily_data function)
+            kmin:       minimum number of clusters
+            kmax:       maximum number of clusters
+            plot:       flag to determine if an elbow plot should be displayed
+
+        Returns:
+            n_clusters:     the optimal number of clusters for the given data
+            inertia_values: within-cluster sum-of-squares
+        """
+        # Check kmin and kmax for validity
+        if kmin is None:
+            kmin = 4
+        if kmax is None:
+            kmax = 30
+            _logger.warning(f"kmax was not set - using a default value of 30.")
+
+        if not isinstance(kmin, int):
+            raise ValueError(f"kmin must be an integer, but {kmin} is not an integer")
+        if not isinstance(kmax, int):
+            raise ValueError(f"kmax must be an integer, but {kmax} is not an integer")
+        if kmin < 1:
+            raise ValueError(f"kmin must be > 0, but {kmin} is provided.")
+        if kmax < 1:
+            raise ValueError(f"kmax must be > 0, but {kmax} is provided.")
+        if kmin >= kmax:
+            raise ValueError(f"kmin must be less than kmax, but {kmin} >= {kmax}")
+
+        k_values = range(kmin, kmax + 1)
+        inertia_values = []
+
+        np.random.seed(self._seed)
+
+        # Compute the inertia (SSE) for k clusters
+        for k in k_values:
+            kmeans = KMeans(n_clusters=k).fit(daily_data.transpose())
+            inertia_values.append(kmeans.inertia_)
+
+        # Identify the "elbow point"
+        elbow_point = KneeLocator(
+            k_values, inertia_values, curve="convex", direction="decreasing"
+        )
+        n_clusters = elbow_point.knee
+
+        if n_clusters is None:
+            raise ValueError(
+                f"Could not find elbow point for given kmin, kmax. Consider increasing the range of kmin, kmax."
+            )
+
+        print(f"Optimal # of clusters is: {n_clusters}")
+
+        if int(n_clusters) + 2 >= kmax:
+            _logger.warning(
+                f"Optimal number of clusters is close to kmax: {kmax}. Consider increasing kmax."
+            )
+
+        if plot == True:
+            plt.plot(k_values, inertia_values)
```

Comment:
Should this be moved into a separate method so that it can be called by the user separately?

Reply:
Looking more closely, I would recommend breaking this into a separate method. This will make testing easier as well.